nano/node/hinted_scheduler.cpp

```diff
@@ -0,0 +1,122 @@
+#include <nano/node/hinted_scheduler.hpp>
+#include <nano/node/node.hpp>
+
+nano::hinted_scheduler::hinted_scheduler (config const & config_a, nano::node & node_a, nano::vote_cache & inactive_vote_cache_a, nano::active_transactions & active_a, nano::online_reps & online_reps_a) :
+	config_m{ config_a },
+	node{ node_a },
+	inactive_vote_cache{ inactive_vote_cache_a },
+	active{ active_a },
+	online_reps{ online_reps_a },
+	stopped{ false }
+{
+}
+
+nano::hinted_scheduler::~hinted_scheduler ()
+{
+	stop ();
+	if (thread.joinable ()) // Ensure thread was started
+	{
+		thread.join ();
+	}
+}
+
+void nano::hinted_scheduler::start ()
+{
+	debug_assert (!thread.joinable ());
+	thread = std::thread{
+		[this] () { run (); }
+	};
+}
+
+void nano::hinted_scheduler::stop ()
+{
+	nano::unique_lock<nano::mutex> lock{ mutex };
+	stopped = true;
+	notify ();
+}
+
+void nano::hinted_scheduler::notify ()
+{
+	condition.notify_all ();
+}
+
+bool nano::hinted_scheduler::predicate (nano::uint128_t const & minimum_tally) const
+{
+	// Check if there is space inside AEC for a new hinted election
+	if (active.vacancy_hinted () > 0)
+	{
+		// Check if there is any vote cache entry surpassing our minimum vote tally threshold
+		if (inactive_vote_cache.peek (minimum_tally))
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+bool nano::hinted_scheduler::run_one (nano::uint128_t const & minimum_tally)
+{
+	if (auto top = inactive_vote_cache.pop (minimum_tally); top)
+	{
+		const auto hash = top->hash; // Hash of block we want to hint
+
+		// Check if block exists
+		auto block = node.block (hash);
+		if (block != nullptr)
+		{
+			// Ensure block is not already confirmed
+			if (!node.block_confirmed_or_being_confirmed (hash))
```

Comment:
I wonder if this check is better to be somewhere else where multiple blocks can be checked n one transaction.

Reply:
Do you have something concrete in mind? I can't think of a better way, caching block confirmation for multiple blocks earlier is pointless as a block can be already confirmed once it gets to the above line.

Reply:
No, not really, I am not expecting a change. Just wanted to know your thoughts on it.
I was thinking probably inside the AEC but it is OK as it is.